openrdp/scripts/bootscan.py

```diff
@@ -109,48 +116,86 @@ def find_potential_events(self, pair1, pair2):
     def scan(self, i):
         window = self.align[:, i:i + self.win_size]
         # Make bootstrap replicates of alignment
-        dists = []
+        num_arrays = self.num_replicates
+        array_shape = ((self.align.shape[0] - 1) * (self.align.shape[0]) //2,)
+        dists = np.empty((num_arrays,) + array_shape, dtype=np.float16)
         np.random.seed(self.random_seed)
         random.seed(self.random_seed)
         for rep in range(self.num_replicates):
             # Shuffle columns with replacement
             rep_window = window[:, np.random.randint(0, window.shape[1], window.shape[1])]
-            dist_mat = squareform(pdist(rep_window, jc_distance))
-            dists.append(dist_mat)
-        return dists
+            dist_mat = pdist(rep_window, jc_distance)
+            dists[rep] = dist_mat
+
+        with h5py.File('dist_mat_{}.h5'.format(i//self.step_size), 'w') as f:
+            f.create_dataset('dist_mat_{}'.format(i//self.step_size), data=dists)
+
+    def merge_h5py_files(self, src_file, dst_file):
+        with h5py.File(src_file, 'r') as src, h5py.File(dst_file, 'a') as dst:
+            src_items = list(src.items())
+            for name, obj in src_items:
+                dst.create_dataset(name, data=obj[()])
 
     def do_scanning_phase(self, align):
         """
         Perform scanning phase of the Bootscan/Recscan algorithm
         :param align: a n x m array of aligned sequences
         """
-        with multiprocessing.Pool() as p:
-            all_dists = p.map(self.scan, range(0, align.shape[1], self.step_size))
 
-        return all_dists
+        # Remove previous .h5 files
+        h5_files = glob.glob('*.h5')
+        for file in h5_files:
+            if os.path.exists(file):
+                os.remove(file)
+
+        with multiprocessing.Pool(self.np) as p:
```

Comment:
next part of merge conflict - I don't think there are any issues here, this is @GopiGugan 's method of writing data to filesystem before running multiple threads.